File: crates/libs/clawdstrike/src/async_guards/runtime.rs
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;

use dashmap::DashMap;
use futures::stream::{FuturesUnordered, StreamExt};

use crate::async_guards::cache::TtlCache;
use crate::async_guards::circuit_breaker::CircuitBreaker;
use crate::async_guards::http::HttpClient;
use crate::async_guards::rate_limit::TokenBucket;
use crate::async_guards::retry;
use crate::async_guards::types::{
    AsyncGuard, AsyncGuardConfig, AsyncGuardErrorKind, RateLimitConfig,
};
use crate::guards::{GuardAction, GuardContext, GuardResult, Severity};
use crate::policy::{AsyncExecutionMode, TimeoutBehavior};

const DEFAULT_BACKGROUND_IN_FLIGHT_LIMIT: usize = 64;

#[derive(Clone, Debug)]
pub enum OwnedGuardAction {
    FileAccess {
        path: String,
    },
    FileWrite {
        path: String,
        content: Vec<u8>,
    },
    NetworkEgress {
        host: String,
        port: u16,
    },
    ShellCommand {
        commandline: String,
    },
    McpTool {
        tool: String,
        args: serde_json::Value,
    },
    Patch {
        path: String,
        diff: String,
    },
    Custom {
        kind: String,
        data: serde_json::Value,
    },
}

impl OwnedGuardAction {
    pub fn from_borrowed(action: &GuardAction<'_>) -> Self {
        match action {
            GuardAction::FileAccess(p) => Self::FileAccess {
                path: (*p).to_string(),
            },
            GuardAction::FileWrite(p, c) => Self::FileWrite {
                path: (*p).to_string(),
                content: c.to_vec(),
            },
            GuardAction::NetworkEgress(h, p) => Self::NetworkEgress {
                host: (*h).to_string(),
                port: *p,
            },
            GuardAction::ShellCommand(c) => Self::ShellCommand {
                commandline: (*c).to_string(),
            },
            GuardAction::McpTool(t, args) => Self::McpTool {
                tool: (*t).to_string(),
                args: (*args).clone(),
            },
            GuardAction::Patch(p, d) => Self::Patch {
                path: (*p).to_string(),
                diff: (*d).to_string(),
            },
            GuardAction::Custom(kind, data) => Self::Custom {
                kind: (*kind).to_string(),
                data: (*data).clone(),
            },
        }
    }

    pub fn as_guard_action(&self) -> GuardAction<'_> {
        match self {
            Self::FileAccess { path } => GuardAction::FileAccess(path),
            Self::FileWrite { path, content } => GuardAction::FileWrite(path, content),
            Self::NetworkEgress { host, port } => GuardAction::NetworkEgress(host, *port),
            Self::ShellCommand { commandline } => GuardAction::ShellCommand(commandline),
            Self::McpTool { tool, args } => GuardAction::McpTool(tool, args),
            Self::Patch { path, diff } => GuardAction::Patch(path, diff),
            Self::Custom { kind, data } => GuardAction::Custom(kind, data),
        }
    }
}

pub struct AsyncGuardRuntime {
    http: HttpClient,
    caches: DashMap<String, Arc<TtlCache>>,
    limiters: DashMap<String, Arc<TokenBucket>>,
    breakers: DashMap<String, Arc<CircuitBreaker>>,
    background_slots: Arc<tokio::sync::Semaphore>,
    background_in_flight_limit: usize,
    background_running: AtomicUsize,
    background_peak_running: AtomicUsize,
    background_dropped: AtomicUsize,
}

impl Default for AsyncGuardRuntime {
    fn default() -> Self {
        Self::new()
    }
}

impl AsyncGuardRuntime {
    pub fn new() -> Self {
        Self::with_background_in_flight_limit(DEFAULT_BACKGROUND_IN_FLIGHT_LIMIT)
    }

    pub fn with_background_in_flight_limit(limit: usize) -> Self {
        let limit = limit.max(1);
        Self {
            http: HttpClient::new(),
            caches: DashMap::new(),
            limiters: DashMap::new(),
            breakers: DashMap::new(),
            background_slots: Arc::new(tokio::sync::Semaphore::new(limit)),
            background_in_flight_limit: limit,
            background_running: AtomicUsize::new(0),
            background_peak_running: AtomicUsize::new(0),
            background_dropped: AtomicUsize::new(0),
        }
    }

    pub fn http(&self) -> &HttpClient {
        &self.http
    }

    pub fn background_inflight_limit(&self) -> usize {
        self.background_in_flight_limit
    }

    pub fn background_inflight_count(&self) -> usize {
        self.background_running.load(Ordering::Relaxed)
    }

    pub fn background_peak_inflight(&self) -> usize {
        self.background_peak_running.load(Ordering::Relaxed)
    }

    pub fn background_dropped_count(&self) -> usize {
        self.background_dropped.load(Ordering::Relaxed)
    }

    pub async fn evaluate_async_guards(
        self: &Arc<Self>,
        guards: &[Arc<dyn AsyncGuard>],
        action: &GuardAction<'_>,
        context: &GuardContext,
        fail_fast: bool,
    ) -> Vec<GuardResult> {
        let mut out: Vec<GuardResult> = Vec::new();

        let mut sequential: Vec<(usize, Arc<dyn AsyncGuard>)> = Vec::new();
        let mut parallel: Vec<(usize, Arc<dyn AsyncGuard>)> = Vec::new();
        let mut background: Vec<(usize, Arc<dyn AsyncGuard>)> = Vec::new();

        for (idx, g) in guards.iter().cloned().enumerate() {
            if !g.handles(action) {
                continue;
            }

            match g.config().execution_mode {
                AsyncExecutionMode::Sequential => sequential.push((idx, g)),
                AsyncExecutionMode::Parallel => parallel.push((idx, g)),
                AsyncExecutionMode::Background => background.push((idx, g)),
            }
        }

        // Sequential guards (stable order by policy list).
        sequential.sort_by_key(|(idx, _)| *idx);
        for (_idx, g) in sequential {
            let result = self.evaluate_one(g, action, context).await;
            let denied = !result.allowed;
            out.push(result);
            if fail_fast && denied {
                return out;
            }
        }

        // Parallel guards. Short-circuit on deny.
        parallel.sort_by_key(|(idx, _)| *idx);
        if !parallel.is_empty() {
            let results_by_idx: DashMap<usize, GuardResult> = DashMap::new();
            let mut futs = FuturesUnordered::new();

            for (idx, g) in parallel.iter().cloned() {
                let runtime = Arc::clone(self);
                futs.push(async move { (idx, runtime.evaluate_one(g, action, context).await) });
            }

            let mut denied = false;
            while let Some((idx, res)) = futs.next().await {
                if !res.allowed {
                    denied = true;
                }
                results_by_idx.insert(idx, res);
                if fail_fast && denied {
                    break;
                }
            }

            if fail_fast && denied {
                // Remaining futures are dropped here (best-effort cancellation).
                drop(futs);
            }

            for (idx, g) in parallel {
                if let Some((_, res)) = results_by_idx.remove(&idx) {
                    let denied = !res.allowed;
                    out.push(res);
                    if fail_fast && denied {
                        return out;
                    }
                } else {
                    // Canceled due to deny in another parallel guard.
                    out.push(
                        GuardResult::warn(
                            g.name(),
                            "Canceled due to earlier deny in parallel group",
                        )
                        .with_details(serde_json::json!({ "canceled": true })),
                    );
                }
            }

            if fail_fast && denied {
                return out;
            }
        }

        // Background guards: schedule and return allow placeholders.
        if !background.is_empty() {
            let owned_action = OwnedGuardAction::from_borrowed(action);
            let ctx = context.clone();

            background.sort_by_key(|(idx, _)| *idx);
            for (_idx, g) in background {
                if self.spawn_background(g.clone(), owned_action.clone(), ctx.clone()) {
                    out.push(
                        GuardResult::allow(g.name()).with_details(serde_json::json!({
                            "background": true,
                            "note": "scheduled",
                            "in_flight_limit": self.background_inflight_limit()
                        })),
                    );
                } else {
                    out.push(
                        GuardResult::warn(
                            g.name(),
                            "background guard dropped due to in-flight limit",
                        )
                        .with_details(serde_json::json!({
                            "background": true,
                            "note": "dropped",
                            "in_flight_limit": self.background_inflight_limit(),
                            "dropped_total": self.background_dropped_count()
                        })),
                    );
                }
            }
        }

        out
    }

    fn spawn_background(
        self: &Arc<Self>,
        guard: Arc<dyn AsyncGuard>,
        action: OwnedGuardAction,
        context: GuardContext,
    ) -> bool {
        let permit = match self.background_slots.clone().try_acquire_owned() {
            Ok(permit) => permit,
            Err(_) => {
                self.background_dropped.fetch_add(1, Ordering::Relaxed);
                return false;
            }
        };

        let runtime = Arc::clone(self);
        tokio::spawn(async move {
            let running = runtime.background_running.fetch_add(1, Ordering::Relaxed) + 1;
            update_max(&runtime.background_peak_running, running);

            let borrowed = action.as_guard_action();
            let result = runtime
                .evaluate_one(guard.clone(), &borrowed, &context)
                .await;

            if !result.allowed {
                tracing::warn!(
                    guard = guard.name(),
                    action = "alert",
                    message = %result.message,
                    severity = ?result.severity,
                    "background async guard would have denied"
                );
            }

            runtime.background_running.fetch_sub(1, Ordering::Relaxed);
            drop(permit);
        });
        true
    }

    async fn evaluate_one(
        self: &Arc<Self>,
        guard: Arc<dyn AsyncGuard>,
        action: &GuardAction<'_>,
        context: &GuardContext,
    ) -> GuardResult {
        let name = guard.name().to_string();
        let cfg = guard.config().clone();

        let cache_key = guard.cache_key(action, context);
        if cfg.cache_enabled {
            if let Some(ref key) = cache_key {
                if let Some(mut cached) = self.cache_for(&name, &cfg).get_guard_result(key) {
                    let merged =
                        merge_details(cached.details.take(), serde_json::json!({ "cache": "hit" }));
                    cached.details = Some(merged);
                    return cached;
                }
            }
        }

        if let Some(cb_cfg) = cfg.circuit_breaker.clone() {
            let breaker = self
                .breakers
                .entry(name.clone())
                .or_insert_with(|| Arc::new(CircuitBreaker::new(cb_cfg)))
                .clone();

            if breaker.before_request().await.is_err() {
                return fallback(
                    &name,
                    &cfg,
                    AsyncGuardErrorKind::CircuitOpen,
                    "circuit breaker open",
                    cache_key.as_deref(),
                    self.cache_for(&name, &cfg),
                );
            }
        }

        if let Some(RateLimitConfig {
            requests_per_second,
            burst,
        }) = cfg.rate_limit.clone()
        {
            let limiter = self
                .limiters
                .entry(name.clone())
                .or_insert_with(|| Arc::new(TokenBucket::new(requests_per_second, burst)))
                .clone();
            limiter.acquire().await;
        }

        let attempt = async {
            if let Some(retry_cfg) = cfg.retry.clone() {
                retry::retry(&retry_cfg, |attempt| {
                    let guard = guard.clone();
                    async move {
                        guard
                            .check_uncached(action, context, &self.http)
                            .await
                            .map_err(|mut e| {
                                if e.kind == AsyncGuardErrorKind::Other && e.status.is_none() {
                                    e.message =
                                        format!("attempt {} failed: {}", attempt + 1, e.message);
                                }
                                e
                            })
                    }
                })
                .await
            } else {
                guard.check_uncached(action, context, &self.http).await
            }
        };

        let timed = tokio::time::timeout(cfg.timeout, attempt).await;

        match timed {
            Ok(Ok(res)) => {
                // Cache success.
                if cfg.cache_enabled {
                    if let Some(ref key) = cache_key {
                        self.cache_for(&name, &cfg).set_guard_result(
                            key.clone(),
                            &res,
                            cfg.cache_ttl,
                        );
                    }
                }

                if cfg.circuit_breaker.is_some() {
                    if let Some(b) = self.breakers.get(&name) {
                        b.record_success().await;
                    }
                }

                res
            }
            Ok(Err(err)) => {
                if cfg.circuit_breaker.is_some() {
                    if let Some(b) = self.breakers.get(&name) {
                        b.record_failure().await;
                    }
                }

                fallback(
                    &name,
                    &cfg,
                    err.kind,
                    &err.message,
                    cache_key.as_deref(),
                    self.cache_for(&name, &cfg),
                )
            }
            Err(_) => {
                if cfg.circuit_breaker.is_some() {
                    if let Some(b) = self.breakers.get(&name) {
                        b.record_failure().await;
                    }
                }

                fallback(
                    &name,
                    &cfg,
                    AsyncGuardErrorKind::Timeout,
                    "timeout",
                    cache_key.as_deref(),
                    self.cache_for(&name, &cfg),
                )
            }
        }
    }

    fn cache_for(&self, guard_name: &str, cfg: &AsyncGuardConfig) -> Arc<TtlCache> {
        self.caches
            .entry(guard_name.to_string())
            .or_insert_with(|| Arc::new(TtlCache::new(cfg.cache_max_size_bytes.max(1024))))
            .clone()
    }
}

fn update_max(target: &AtomicUsize, candidate: usize) {
    loop {
        let current = target.load(Ordering::Relaxed);
        if candidate <= current {
            return;
        }
        if target
            .compare_exchange(current, candidate, Ordering::Relaxed, Ordering::Relaxed)
            .is_ok()
        {
            return;
        }
    }
}

fn fallback(
    guard: &str,
    cfg: &AsyncGuardConfig,
    kind: AsyncGuardErrorKind,
    message: &str,
    cache_key: Option<&str>,
    cache: Arc<TtlCache>,
) -> GuardResult {
    let detail = serde_json::json!({
        "async_error": {
            "kind": format!("{:?}", kind),
            "message": message,
        }
    });

    match cfg.on_timeout {
        TimeoutBehavior::Allow => GuardResult::allow(guard).with_details(detail),
        TimeoutBehavior::Warn => {
            GuardResult::warn(guard, format!("Async guard error: {}", message)).with_details(detail)
        }
        TimeoutBehavior::Deny => GuardResult::block(
            guard,
            Severity::Error,
            format!("Async guard error: {}", message),
        )
        .with_details(detail),
        TimeoutBehavior::Defer => {
            if let Some(key) = cache_key {
                if let Some(mut cached) = cache.get_guard_result(key) {
                    cached.details = Some(merge_details(
                        cached.details,
                        serde_json::json!({ "cache": "defer_hit" }),
                    ));
                    return cached;
                }
            }
            GuardResult::warn(guard, "Deferred async guard had no cached result")
                .with_details(detail)
        }
    }
}

fn merge_details(
    existing: Option<serde_json::Value>,
    extra: serde_json::Value,
) -> serde_json::Value {
    match existing {
        None => extra,
        Some(mut a) => match extra {
            serde_json::Value::Object(b_obj) => {
                if let serde_json::Value::Object(ref mut a_obj) = a {
                    for (k, v) in b_obj {
                        a_obj.insert(k, v);
                    }
                    a
                } else {
                    serde_json::json!({ "details": a, "extra": serde_json::Value::Object(b_obj) })
                }
            }
            b => serde_json::json!({ "details": a, "extra": b }),
        },
    }
}

File: packages/policy/clawdstrike-policy/src/async/runtime.ts
import type { PolicyEvent } from '@clawdstrike/adapter-core';

import { GuardCache } from './cache.js';
import { CircuitBreaker } from './circuit-breaker.js';
import { FetchHttpClient } from './http.js';
import { TokenBucket } from './rate-limit.js';
import { retry } from './retry.js';
import type { AsyncGuard, AsyncGuardConfig, GuardResult, HttpClient } from './types.js';

export class AsyncGuardRuntime {
  private readonly http: HttpClient;
  private readonly caches = new Map<string, GuardCache>();
  private readonly limiters = new Map<string, TokenBucket>();
  private readonly breakers = new Map<string, CircuitBreaker>();

  constructor(http: HttpClient = new FetchHttpClient()) {
    this.http = http;
  }

  async evaluateAsyncGuards(
    guards: Array<{ index: number; guard: AsyncGuard }>,
    event: PolicyEvent,
  ): Promise<GuardResult[]> {
    const out: GuardResult[] = [];

    const sequential: Array<{ index: number; guard: AsyncGuard }> = [];
    const parallel: Array<{ index: number; guard: AsyncGuard }> = [];
    const background: Array<{ index: number; guard: AsyncGuard }> = [];

    for (const g of guards) {
      if (!g.guard.handles(event)) continue;
      if (g.guard.config.executionMode === 'sequential') sequential.push(g);
      else if (g.guard.config.executionMode === 'parallel') parallel.push(g);
      else background.push(g);
    }

    sequential.sort((a, b) => a.index - b.index);
    for (const g of sequential) {
      const res = await this.evaluateOne(g.guard, event);
      out.push(res);
      if (!res.allowed) return out;
    }

    parallel.sort((a, b) => a.index - b.index);
    if (parallel.length > 0) {
      const byIndex = new Map<number, GuardResult>();
      const controller = new AbortController();
      let denied = false;

      const pending = new Map<number, Promise<{ index: number; res: GuardResult }>>();
      for (const g of parallel) {
        const p = this.evaluateOne(g.guard, event, controller.signal).then((res) => ({
          index: g.index,
          res,
        }));
        pending.set(g.index, p);
      }

      while (pending.size > 0) {
        const { index, res } = await Promise.race(pending.values());
        pending.delete(index);
        byIndex.set(index, res);
        if (!res.allowed) {
          denied = true;
          controller.abort();
          break;
        }
      }

      for (const g of parallel) {
        const res = byIndex.get(g.index);
        if (res) {
          out.push(res);
          if (!res.allowed) return out;
        } else {
          out.push({
            allowed: true,
            guard: g.guard.name,
            severity: 'medium',
            message: 'Canceled due to earlier deny in parallel group',
            details: { canceled: true },
          });
        }
      }

      if (denied) return out;
    }

    background.sort((a, b) => a.index - b.index);
    if (background.length > 0) {
      for (const g of background) {
        void this.evaluateOne(g.guard, event).then((res) => {
          if (!res.allowed) {
            // Background mode never changes the immediate decision.
            // Emit an alert-style log for visibility.
            // eslint-disable-next-line no-console
            console.warn(
              `[clawdstrike] background async guard would have denied: ${g.guard.name}: ${res.message}`,
            );
          }
        });

        out.push({
          allowed: true,
          guard: g.guard.name,
          severity: 'low',
          message: 'Allowed',
          details: { background: true, note: 'scheduled' },
        });
      }
    }

    return out;
  }

  private async evaluateOne(guard: AsyncGuard, event: PolicyEvent, signal?: AbortSignal): Promise<GuardResult> {
    const cfg = guard.config;

    const cacheKey = guard.cacheKey(event);
    if (cacheKey && cfg.cacheEnabled) {
      const cached = this.cacheFor(guard.name, cfg).get(cacheKey);
      if (cached) {
        return withMergedDetails(cached, { cache: 'hit' });
      }
    }

    if (cfg.circuitBreaker) {
      const breaker = this.breakerFor(guard.name, cfg);
      const ok = breaker.beforeRequest();
      if (!ok.ok) {
        return this.fallback(guard, cfg, 'circuit_open', 'circuit breaker open', cacheKey);
      }
    }

    if (cfg.rateLimit) {
      await this.limiterFor(guard.name, cfg).acquire(signal);
    }

    const controller = new AbortController();
    const combined = signal ? anySignal([signal, controller.signal]) : controller;

    const timeoutId = setTimeout(() => controller.abort(), cfg.timeoutMs);
    timeoutId.unref?.();

    try {
      const attempt = async () => {
        if (cfg.retry) {
          return await retry(
            cfg.retry,
            async () => await guard.checkUncached(event, this.http, combined.signal),
            combined.signal,
          );
        }
        return await guard.checkUncached(event, this.http, combined.signal);
      };

      const res = await attempt();

      if (cacheKey && cfg.cacheEnabled) {
        this.cacheFor(guard.name, cfg).set(cacheKey, res);
      }

      if (cfg.circuitBreaker) {
        this.breakerFor(guard.name, cfg).recordSuccess();
      }

      return res;
    } catch (err) {
      if (cfg.circuitBreaker) {
        this.breakerFor(guard.name, cfg).recordFailure();
      }

      const message = err instanceof Error ? err.message : String(err);
      const kind = combined.signal.aborted ? 'timeout' : 'other';
      return this.fallback(guard, cfg, kind, message, cacheKey);
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private cacheFor(guardName: string, cfg: AsyncGuardConfig): GuardCache {
    const existing = this.caches.get(guardName);
    if (existing) return existing;
    const cache = new GuardCache(cfg.cacheMaxSizeBytes, cfg.cacheTtlSeconds);
    this.caches.set(guardName, cache);
    return cache;
  }

  private limiterFor(guardName: string, cfg: AsyncGuardConfig): TokenBucket {
    const existing = this.limiters.get(guardName);
    if (existing) return existing;
    const rl = cfg.rateLimit ?? { requestsPerSecond: 0, burst: 1 };
    const limiter = new TokenBucket(rl.requestsPerSecond, Math.max(1, rl.burst));
    this.limiters.set(guardName, limiter);
    return limiter;
  }

  private breakerFor(guardName: string, cfg: AsyncGuardConfig): CircuitBreaker {
    const existing = this.breakers.get(guardName);
    if (existing) return existing;
    const cb = cfg.circuitBreaker ?? { failureThreshold: 5, resetTimeoutMs: 30_000, successThreshold: 2 };
    const breaker = new CircuitBreaker(cb.failureThreshold, cb.resetTimeoutMs, cb.successThreshold);
    this.breakers.set(guardName, breaker);
    return breaker;
  }

  private fallback(
    guard: AsyncGuard,
    cfg: AsyncGuardConfig,
    kind: string,
    message: string,
    cacheKey: string | null,
  ): GuardResult {
    const details = { async_error: { kind, message } };

    if (cfg.onTimeout === 'allow') {
      return { allowed: true, guard: guard.name, severity: 'low', message: 'Allowed', details };
    }

    if (cfg.onTimeout === 'deny') {
      return { allowed: false, guard: guard.name, severity: 'high', message: `Async guard error: ${message}`, details };
    }

    if (cfg.onTimeout === 'defer') {
      if (cacheKey) {
        const cached = this.cacheFor(guard.name, cfg).get(cacheKey);
        if (cached) return withMergedDetails(cached, { cache: 'defer_hit' });
      }
      return {
        allowed: true,
        guard: guard.name,
        severity: 'medium',
        message: 'Deferred async guard had no cached result',
        details,
      };
    }

    // warn (default)
    return { allowed: true, guard: guard.name, severity: 'medium', message: `Async guard error: ${message}`, details };
  }
}

function withMergedDetails(result: GuardResult, extra: Record<string, unknown>): GuardResult {
  const details = result.details ? { ...result.details, ...extra } : extra;
  return { ...result, details };
}

function anySignal(signals: AbortSignal[]): AbortController {
  const controller = new AbortController();
  const onAbort = () => controller.abort();
  for (const s of signals) {
    if (s.aborted) {
      controller.abort();
      return controller;
    }
    s.addEventListener('abort', onAbort, { once: true });
  }
  return controller;
}
